### DISCUSSION TITLE
Alpine with Echarts

### DISCUSSION THREAD

**Discussion by mccarlosen** [alpinejs/alpine#1953] [7. Bugs (archived)] [answered]
I cannot determine the error. But it does not allow the correct functioning of the echarts graph. When you interact with the legend it gives an error.

![ezgif com-gif-maker (1)](https://user-images.githubusercontent.com/8524361/129579482-f3e0280f-5707-4965-a088-78c44065fef3.gif)

However when using vanilla js it works fine.

![ezgif com-gif-maker (2)](https://user-images.githubusercontent.com/8524361/129579616-eeec12d5-8700-4c94-af6f-38885b0f59f5.gif)

Here are other examples:

[Alpine-with-echarts-1](https://codesandbox.io/s/alpinejs-with-echarts-1-6jls0)
[Alpine-with-echarts-2](https://codesandbox.io/s/alpinejs-with-echarts-2-f8lwx)
[Echarts-without-Alpine](https://codesandbox.io/s/echarts-with-vanilla-js-7lqw6)

Any ideas I appreciate.

**Comment by KevinBatdorf** [comment, marked as answer]
You could change it to something like this:
```js
  init() {
    window.echart = echarts.init(this.$el);
    window.echart.setOption(this.option);
  }
```

**Reply by mccarlosen** [reply]
Awesome @KevinBatdorf ... simple solution. Thanks for such a quick response.

**Comment by SimoTod** [comment]
You can also define a variable outside the reactive part (so it's still encapsulated in your component and it won't conflict if you have multiple graphs on the same page)

```javascript
import * as echarts from "echarts";
import Alpine from "alpinejs";

window.Alpine = Alpine;

Alpine.data("echart", () => {
  let echart;
  {
    option: {
      ...
    },
    init() {
      echart = echarts.init(this.$el);
      echart.setOption(this.option);
    }
  }
);

Alpine.start();
```

**Reply by KevinBatdorf** [reply]
This is the more maintainable approach. I'll start recommending this. 

Mainly because it would force you to interact with the api from within the same scope and not outside directly on the global scope. So even if the application only has one chart, this is a better practice.

(you forgot the return statement though)

**Reply by mccarlosen** [reply]
@KevinBatdorf @SimoTod It is definitely a good solution too. However, I have a question as to how I am approaching this. I have created a livewire component and load the data in this way.

```html
<div class="echarts"
x-data="{
  options: {
    title: {
      text: '{{ $title }}'
    },
    tooltip: {{ $tooltip }},
    legend: {{ $legend }},
    backgroundColor: '{{ $backgroundColor }}',
    grid: {{ $grid }},
    xAxis: {{ $xAxis }},
    yAxis: {{ $yAxis }},
    series: {{ $series }}
  },
  resizeHandler: $store.utils.$debounce(function() {
    var echart = echarts.getInstanceByDom(this);
    echart.resize();
  }, 100, { leading: true }),
  init() {
    window.echart = echarts.init($el);
    window.echart.setOption(this.options);
    $store.utils.$addListener($el, this.resizeHandler);
    window.addEventListener('resize',function(){
      window.echart = echarts.init($el);
      window.echart.resize()
    })
  }
}">

</div>
```
```html
<livewire:components.echarts 
  :identify="$id" 
  :grid="$grid" 
  :xAxis="$xAxisData" 
  :yAxis="$yAxisData" 
  :series="$seriesRating" 
  :legend="$legend" 
  :tooltip="$tooltip" 
  :key="'chart'.$id" 
/>
```
![ezgif com-gif-maker (4)](https://user-images.githubusercontent.com/8524361/129988920-07a703f2-4cb7-4265-900e-7665aadd43d5.gif)

In this case, is the JS module option better? How could I get the data using the JS module option without having to use fetch etc?

I'm just raising it to see if anyone has a better way to approach it (Alpine and Livewire).

**Reply by KevinBatdorf** [reply]
You would need to extract it to an outside `Alpine.data` method to set it up the way SimoTod recommended. There's probably a way to declare it inside `x-data` but maybe that's not a good idea.

When you use `window.something` you're adding that to the global scope. That's not usually a big deal, but you risk overriding it or if your app gets complex, you risk accessing it in multiple ways which could cause bugs. Having two components on the page wouldn't work how you expect either as the second one would override the first (is that not happening?).

```js
init() {
  window.echart = echarts.init($el);
  window.echart.setOption(this.options);
  $store.utils.$addListener($el, this.resizeHandler);
  window.addEventListener('resize',function(){
    window.echart = echarts.init($el);
    window.echart.resize()
  })
}
```

The code above has a few issues (unrelated to your initial question):
- When you init the echart library, it's likely adding event listeners somewhere, then adding them again on every resize
- You probably don't need the echart `init` inside the resize? if you do, look if they have a `destroy` method to remove any event listeners. And possibly "debounce" it so it's not running every resize event (it fires a lot)
- `window.addEventListener` is adding this event to window in a way Alpine wont clean up in the case the component is removed. You could do `@resize.window="echart.resize()"` Also not critical if you aren't removing the component, but that's the "Alpine way" - But maybe you are given it's a livewire component (I'm not super familiar with LW)?

**Reply by mccarlosen** [reply]
@KevinBatdorf Thanks for the recommendations and corrections. You are right every time the component is updated, `init()` is executed. As for if the first chart overrides the second chart, it doesn't happen. Both work as expected.

**Reply by mccarlosen** [reply]
I think it looks better this way:

```html
<div class="echarts"
x-data="{
  options: {
    ...
  },
  resizeHandler: $store.utils.$debounce(function() {
    var echart = echarts.getInstanceByDom(this.$el);
    echart.resize();
  }, 100, { leading: true }),
  init() {
    window.echart = echarts.init($el);
    window.echart.setOption(this.options);
  }
}" @resize.window="resizeHandler()">
</div>
```
So far I can't find a flexible way to extract the `Alpine.data` method for a Livewire component. I keep investigating. Thank you very much for your help.

**Reply by SimoTod** [reply]
If you want to inline x-data and you are not bothered about accessing the echart variable directly outside init (I see you can get it back using echarts.getInstanceByDom anyway), you can do 
```javascript
  init() {
    let echart = echarts.init($el);
    echart.setOption(this.options);
  }
```

By the way, putting that on window is not a problem, the main thing to remember as Kevin said is that if you have 2 components using the same variable and after you try to access window.echart wanting the first graph, you'll get the second one.
This won't be a problem with your current code since you do `var echart = echarts.getInstanceByDom(this.$el); echart.resize();` but if you had `window.echart.resize();` and 2 graphs instead it wouldn't work correctly for the first one.

**Reply by mccarlosen** [reply]
@SimoTod If it makes a lot of sense. Thanks.

**Reply by carlos-verdes** [reply]
Just to share what it worked for me based on all previous answers:
```js
Alpine.data("chart", () => {
  let chart;
  return {
    init() {
      chart = echarts.init(this.$el);

     // init logic here
    },
    resize: {
      ["@resize.window.debounce"]() {
        if (chart) {
          chart.resize();
        }
      }
    }
  }
});
```

Then to use the chart with resize you just need to `x-bind` resize:
```html
<div x-data="chart" x-bind="resize">
...
```
